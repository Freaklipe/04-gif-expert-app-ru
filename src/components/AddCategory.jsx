import { useState } from 'react';


export const AddCategory = ( {/*setCategories, */ onNewCategory} ) => {

    const [inputValue, setInputValue] = useState('');

    // se encarga se recibir lo tipeado en el input, como extra tiene doble destructuracion
    const onIputChange = ( { target:{value} } ) => {
        // console.log(value);
        setInputValue( value );
    };

    // pasa el valor del input a las categorias
    const onSubmit = ( event ) => {
        // previene los auto refresh del formulario
        event.preventDefault();

        // limpieza del string ingresado
        const cleanInputValue = inputValue.trim();

        // valida que minimo exista  1 letra para input
        if ( cleanInputValue.length <= 0) return;

        // pasa el valor del input a las categorias
        // setCategories( categories => [ inputValue, ...categories ] );
        onNewCategory( cleanInputValue );

        // para limpiar el input
        setInputValue('');
    };

  return (
    <form onSubmit={ onSubmit } >
        <input 
            type="text" 
            placeholder="Buscar gifs"
            value={ inputValue }
            onChange={ onIputChange }
        />
    </form>
  )
}
