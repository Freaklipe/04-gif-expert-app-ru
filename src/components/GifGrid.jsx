import PropTypes from 'prop-types';

import { useFetchGifs } from "../hook/useFetchGifs";
import { GifItem } from "./GifItem";


export const GifGrid = ( {category} ) => {


    //////////////////// Resumido en un custom hook///////////////////

    // const [images, setImages] = useState([]);

    // // Simplifica el uso del async para no tener que meterlo como tal dentro del useeffect
    // const getImages = async() => {
    //     // Obtienes las images nuevas
    //     const newImages = await getGifs( category );
    //     // setea las imagenes nuevas
    //     setImages( newImages );
    // }; 

    // // Controla la ejecucion del metodo getImages, para evitar su sobre llamado
    // useEffect(() => {
    //     getImages();
    // }, [category]);

    ///////////////////////////////////////////////////////////////////

    // Custom Hook
    const { images, isLoading } = useFetchGifs( category );
    // console.log(isLoading);

    return (
        <>
            { isLoading && <h2>Cargando...</h2> }
            <h3>{ category }</h3>
            <div className="card-grid">
                {
                    images.map( ( image ) => (
                        <GifItem
                            
                            key={ image.id }
                            // Recordar esta tecnica, es muy comoda
                            { ...image }
                        />
                    ))
                }
            </div>
        </>
    )

}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
