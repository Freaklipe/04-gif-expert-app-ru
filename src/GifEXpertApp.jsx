import { useState } from 'react';

import { AddCategory, GifGrid } from './components';
import './styles/styles.css';


export const GifEXpertApp = () => {

  const [categories, setCategories] = useState([]);

  const onAddCategory = (newCategory) => { 
    // valida si existe y no lo ingresa, la validacion es completa en lower case, para evitar el case sensitibe
    const catLower = categories.map(cat => cat.toLowerCase());
    const newCatLower = newCategory.toLowerCase();
    if( catLower.includes( newCatLower )) return;

    setCategories( [ newCategory, ...categories ] );
    // console.log(newCategory);
    // setCategories( cat => [ newCategory, ...cat ] );
  };

  return (
    <>
        {/* titulo */}
        <h1>@_@ Gif EXpert App *_*</h1>

        {/* input */}
        <AddCategory 
          // setCategories={ setCategories }
          // la implementacion de abajo es mejor, ya que hace llegar el valor y evita complicar las cosas
          onNewCategory={ onAddCategory }
        />

        {/* Listado de gif */}
        
        { 
          categories.map( category =>
            <GifGrid 
              category={ category }
              key={ category } 
            />
          )
        }
        {/* gif item */}

    </>
  );
};
