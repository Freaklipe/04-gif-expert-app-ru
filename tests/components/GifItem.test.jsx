import { render, screen } from '@testing-library/react';
import { GifItem } from '../../src/components';


describe('Pruebas en GiftItem', () => {  

    const title = 'One Piece';
    const url   = 'https://one-piece.com/onepiece.jpg'

    test('debe ser igual al snapshot', () => {  

        const { container } = render( <GifItem title={ title } url={ url } />);
        expect( container ).toMatchSnapshot();
    });

    test('should debe de mostar la imagen con el URL y al ALT indicado', () => {  

        render( <GifItem title={ title } url={ url } />);
        // asi se puede ver como se esta viendo el componente
        // screen.debug();
        // Evalua que la url (propiedades) de la img (elemento) sea la misma que se esta mandando en esta prueba
        // expect( screen.getByRole('img').src ).toBe( url );
        // forma optima de preguntar, cuando has mas de 1 propiedad
        const { src, alt } = screen.getByRole('img');
        expect( src ).toBe( url );
        expect( alt ).toBe( alt );
    });

    test('debe se mostrar el titulo en el componente', () => {  
        render( <GifItem title={ title } url={ url } />);
        // BUsca que exista el titulo dentro del html
        expect( screen.getByText( title) ).toBeTruthy();
    });
});