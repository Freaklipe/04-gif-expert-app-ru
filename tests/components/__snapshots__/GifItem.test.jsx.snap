// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GiftItem debe ser igual al snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="One Piece"
      src="https://one-piece.com/onepiece.jpg"
    />
    <p>
       
      One Piece
       
    </p>
  </div>
</div>
`;
